pkg/spdk/replica.go

```diff
@@ -649,53 +689,376 @@ func (r *Replica) SnapshotDelete(spdkClient *spdkclient.Client, snapshotName str
+	// }
+	// r.rebuildingLvol = &Lvol{
+	// 	Name:     lvolName,
+	// 	UUID:     bdevLvolList[0].UUID,
+	// 	Alias:    bdevLvolList[0].Aliases[0],
+	// 	SpecSize: r.SpecSize,
+	// 	Children: map[string]*Lvol{},
+	// }
+	// if exposeRequired {
+	// 	r.rebuildingPort, _, err = r.portAllocator.AllocateRange(1)
+	// 	if err != nil {
+	// 		return "", err
+	// 	}
+	// 	if err := spdkClient.StartExposeBdev(helpertypes.GetNQN(lvolName), r.rebuildingLvol.UUID, r.IP, strconv.Itoa(int(r.rebuildingPort))); err != nil {
+	// 		return "", err
+	// 	}
+	// }
+
+	r.isRebuilding = true
+	// TODO: Currently, the online rebuilding related APIs are not ready, we use the exposed head lvol for snapshot lvol rebuilding
+	// TODO: Will use r.portAllocator rather than the reserved first port for rebuilding in the future version
+	r.rebuildingLvol = r.ActiveChain[r.ChainLength-1]
+	if exposeRequired {
+		if !r.IsExposed {
+			if err := spdkClient.StartExposeBdev(helpertypes.GetNQN(r.rebuildingLvol.Name), r.rebuildingLvol.UUID, r.IP, strconv.Itoa(int(r.PortStart))); err != nil {
+				return "", err
+			}
+			r.IsExposed = true
+		}
+		r.rebuildingPort = r.PortStart
+	}
+
+	return net.JoinHostPort(r.IP, strconv.Itoa(int(r.rebuildingPort))), nil
+}
+
+func (r *Replica) RebuildingDstFinish(spdkClient *spdkclient.Client, unexposeRequired bool) (err error) {
+	updateRequired := false
+
+	r.Lock()
+	defer func() {
+		r.Unlock()
+
+		if updateRequired {
+			r.UpdateCh <- nil
+		}
+	}()
+
+	if r.State != types.InstanceStateRunning {
+		return fmt.Errorf("invalid state %v for replica %s rebuilding finish", r.State, r.Name)
+	}
+	if !r.isRebuilding {
+		return fmt.Errorf("replica %s is not in rebuilding", r.Name)
+	}
+	if r.rebuildingLvol == nil {
+		return fmt.Errorf("cannot find rebuilding lvol for replica %s rebuilding finish", r.Name)
+	}
+
+	defer func() {
+		if err != nil && r.State != types.InstanceStateError {
+			r.State = types.InstanceStateError
+			updateRequired = true
+		}
+		r.isRebuilding = false
+	}()
+
+	// TODO: For the online rebuilding, Remove the temporary rebuilding lvol and release ports for r.portAllocator
+	// if err := r.rebuildingDstCleanup(spdkClient); err != nil {
+	// 	return err
+	// }
+
+	// For the current offline rebuilding, the temporary rebuilding lvol is actually the head lvol hence there is no need to remove it
+	if unexposeRequired {
+		if err := spdkClient.StopExposeBdev(helpertypes.GetNQN(r.rebuildingLvol.Name)); err != nil && !jsonrpc.IsJSONRPCRespErrorNoSuchDevice(err) {
+			return err
+		}
+		r.IsExposed = false
+	}
+	r.rebuildingPort = 0
+	r.rebuildingLvol = nil
+
+	// TODO: For online rebuilding, connect the snapshot tree/chain with the head lvol then mark the head lvol as mode RW
+
+	// Then re-construct the replica ActiveChain (and SnapshotMap)
+	// TODO: Can we somehow rely on the periodic sync to re-construct the chain?
```

Comment:
If we accept eventual consistency, it should be feasible and it will be good to have one place to do the chain construction. 

However, the primary concern would be consistency like rebuilding has been finished, but the chain is not updated yet?

Reply:
Good point! Then it's better to construct the chain separately, before the rebuilding related fields are cleaned up.
I will remove this comment